### PR TITLE
feat: user data server routes

### PR DESCRIPTION
**Pull request by AasmundN**
# Changes
- `user` server middleware. Fetches the users data from the _db_ and merges it with the _decodedToken_ on the _h3 event context_. The user generated user object is placed on the event context.
- Server plugin `auth.server.ts` to populate `authStore` on initial server render. Runs after the `auth` and `user` server middlewares and only on the initial server render. The plugin has access to both the event context and the _nuxtApp_ instance, which means it can populate the Pinia store. The global `auth` route middleware has been updated accordingly.
- Add `db` export from `firebase` util.
- `hasAccessLevel` server util to check a users accessLevel. 

### API endoints
**`POST /api/user`** endpoint creates or updates a user in the firebase realtime database. A user object in the database has the following format.
```ts
{
  accessLevel?: string[]
  studyProgram: string
  updated: number
}
```
The user entries use the firebase token subject scope `decodedToken.sub` as their object keys. The currently available accessLevels are `'admin'`. A default user has no accesslevel array defined. The updated field holds the date of the last time the user object was updated. The endpoint accepts a request body of the following format.

```ts
{
   sub: string
   accessLevel: string[]
   studyProgram: string
}
```
These specify how the data of the user with _uid_ `sub` is to be updated. Only admins can update other users than their own.

**`GET /api/user/`** endpoint returns the users data object, stored on the event context (see `user` middleware). If the query paramter `scope` is present and has the value `'all'`, all users will be fetched. Only admins can fetch all users.

**`DELETE /api/user/`** endpoint deletes the users data from the database. The `sub` field of the request body specifies which user to remove. Only admins can remove other users than their own.

# Further work
This PR only adds the endpoints to modify the data in the database. The firebase auth composable needs to be updated to use these endpoints, as well as update the users in firebase auth accordingly.

# Resources
- Firebase admin SDK realtime database [reference](https://firebase.google.com/docs/reference/node/firebase.database).

### CODE REVIEW

File: pages/index.vue [hunk 1]
```diff
@@ -46,7 +46,7 @@
 
   const deleteUser = async () => {
     try {
-      const data = await $fetch('/api/user', {
+      await $fetch('/api/user', {
         method: 'DELETE',
         body: {
           sub: 'mogus',
```
Comment (Bissas) [inline pages/index.vue:52]: sus

File: server/api/user/index.post.ts [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  // add or update user in database
+  usersRef.child(sub).set({
+    accessLevel,
+    studyProgram,
+    updated: Date.now(),
+  })
+
+  // user successfully modified
+  sendNoContent(event, 201)
```
Comment (ipeglin) [inline server/api/user/index.post.ts:40]: La oss bare sette alt til 200 OK så ingen skjønner contexten av noen ting 🥴 Neida, looks good

File: server/api/user/index.ts [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+  // user is not authenticated
+  if (!decodedToken)
+    throw createError({
+      statusCode: 401,
+      statusMessage: 'User not authenticated',
+    })
+
+  if (!user)
+    throw createError({
+      statusCode: 404,
```
Comment (ipeglin) [inline server/api/user/index.ts:16]: I `index.delete.ts` thrower du `401` når `!user` er truthy

File: server/api/user/index.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+      statusMessage: 'User data not found',
+    })
+
+  // get scope from query params
+  const { scope } = getQuery(event)
```
Comment (ipeglin) [inline server/api/user/index.ts:21]: Samme som med readBody. Er disse built-in? Er ikke vant med å ha så beleilige composables tilgjengelig "out of the box"

File: server/api/user/index.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+  // reference to users
+  const usersRef = db.ref('users')
+
+  // get all users from db
+  const snapshot = await usersRef.once('value')
+  const dbUsers = snapshot.val()
+
+  // get the first 1000 users from firebase auth
+  const firebaseUsers = await auth.listUsers()
```
Comment (ipeglin) [inline server/api/user/index.ts:34]: Hva er forskjellen på `firebaseUser` og `dbUser`?

Comment (ipeglin) [reply server/api/user/index.ts:34]: Er det snakk om auth vs. bruker dataen?

File: server/api/user/index.delete.ts [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+  // get request body
+  let { sub } = await readBody(event)
+
+  // only admins can remove other users
+  if (!hasAccessLevel(user, 'admin') || !sub) {
+    sub = user.sub
+  }
+
+  // reference to users
+  const usersRef = db.ref('users')
```
[the remark this answers is not in the document]
Comment (AasmundN) [reply server/api/user/index.delete.ts:23]: Jepp, alt fra server/utils autoimporteres i server-filene.

File: server/api/user/index.delete.ts [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+
+  // user is not authenticated
+  if (!user)
+    throw createError({
+      statusCode: 401,
+      statusMessage: 'User not authenticated',
+    })
+
+  // get request body
+  let { sub } = await readBody(event)
```
[the remark this answers is not in the document]
Comment (AasmundN) [reply server/api/user/index.delete.ts:15]: Sub står for subject. Navnet kommer fra OIDC. Men vi kan egentlig bruke UID istedet, bytter.

File: server/api/user/index.post.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
+// POST /api/user
+// endpoint for creating or updating a user in the database
+
+export default defineEventHandler(async (event) => {
+  const { decodedToken, user } = event.context
+
+  // user is not authenticated
+  if (!decodedToken)
```
[the remark this answers is not in the document]
Comment (AasmundN) [reply server/api/user/index.post.ts:8]: Ja. Om decodedToken finnes på context-objektet så vet vi at brukeren er registrert i firebase auth, og dermed kan få lov til å opprette en bruker i vår database.

File: server/api/user/index.post.ts [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    throw createError({
+      statusCode: 401,
+      statusMessage: 'User not authenticated',
+    })
+
+  // get request body
+  let { sub, accessLevel, studyProgram } = await readBody(event)
+
+  if (!studyProgram)
```
[the remark this answers is not in the document]
Comment (AasmundN) [reply server/api/user/index.post.ts:17]: Vi skal ikke lagre navn og e-post i vår database, de ligger på tokenet fra firebase.

File: server/api/user/index.post.ts [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+  if (!hasAccessLevel(user, 'admin') || !sub || !accessLevel) {
+    sub = decodedToken.sub
+    accessLevel = []
+  }
+
+  // reference to users
+  const usersRef = db.ref('users')
+
+  // add or update user in database
+  usersRef.child(sub).set({
```
[the remark this answers is not in the document]
Comment (AasmundN) [reply server/api/user/index.post.ts:33]: Om man ikke bruker den så generer firebase en key for oss.

File: server/api/user/index.ts (hunk 3 above)
Comment (AasmundN) [reply server/api/user/index.ts:16]: Ja, det er med vilje.

File: server/api/user/index.ts (hunk 5 above)
Comment (AasmundN) [reply server/api/user/index.ts:34]: Jepp

File: models/User.ts [hunk 11]
```diff
@@ -1,5 +1,11 @@
-export interface User {
+export interface TokenData {
   uid: string
   email: string
   name?: string | null
 }
+
+export interface User extends TokenData {
+  accessLevel?: string[]
```
[the remark this answers is not in the document]
Comment (AasmundN) [reply models/User.ts:8]: Ja, om brukeren endrer på cookien vil ikke lenger tokenet være gyldig. Da vil ikke auth-middleware gå igjennom.

File: server/api/user/index.ts (hunk 4 above)
Comment (ipeglin) [reply server/api/user/index.ts:21]: Det må jeg sjekke ut. Har ikke noe særlig erfaring med context og slikt

**Pull request merged**